gemini/sample-apps/e2e-gen-ai-app-starter-pack/app/patterns/agent_builder_search/chain.py

```diff
@@ -0,0 +1,159 @@
+# Copyright 2024 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# mypy: disable-error-code="arg-type,attr-defined"
+# pylint: disable=W0613, W0622
+
+import os
+import logging
+from typing import Any, Dict, Iterator, List
+
+from app.patterns.custom_rag_qa.templates import (
+    inspect_conversation_template,
+    rag_template,
+    template_docs,
+)
+from app.utils.output_types import OnChatModelStreamEvent, OnToolEndEvent, custom_chain
+import google
+from langchain.schema import Document
+from langchain.tools import tool
+from langchain_core.messages import ToolMessage
+from langchain_google_community import VertexAISearchRetriever
+from langchain_google_community.vertex_rank import VertexAIRank
+from langchain_google_vertexai import ChatVertexAI, VertexAIEmbeddings
+import vertexai
+
+# Configuration
+EMBEDDING_MODEL = "text-embedding-004"
+LLM_MODEL = "gemini-1.5-flash-002"
+TOP_K = 5
+
+DATASTORE_ID = os.getenv("DATA_STORE_ID", "sample_datastore_id_01")
+VERTEX_AI_LOCATION = os.getenv("SINGLE_REGION", "us-central1")
+AGENT_BUILDER_LOCATION = os.getenv("MULTI_REGION", "us")
+
+# Initialize logging
+logging.basicConfig(level=logging.INFO)
+
+# Initialize Google Cloud and Vertex AI
+credentials, project_id = google.auth.default()
+vertexai.init(project=project_id)
+
+# Set up embedding model and vertex AI retriever
+embedding = VertexAIEmbeddings(model_name=EMBEDDING_MODEL, location=VERTEX_AI_LOCATION)
+
+retriever = VertexAISearchRetriever(
+    project_id=project_id,
+    data_store_id=DATASTORE_ID,
+    location_id=AGENT_BUILDER_LOCATION,
+    engine_data_type=1,
+    # The ratio is set to 0 by default to not use custom
+    # embeddings (We believe the relevance search is more
+    # performant but you can adapt the ratio as you need.
+    custom_embedding_ratio=0,
```

Comment:
maybe add a small liner to describe what this param does as it might not be obvious e.g Controls the ranking of results

Reply:
The docstring of the parameter already contains more details 
i.e. """Controls the ranking of results. Value should be between 0 and 1.
    It will generate the ranking_expression in the following manner:
    "{custom_embedding_ratio} * dotProduct({custom_embedding_field_path}) +
    {1 - custom_embedding_ratio} * relevance_score"
    """
    
    So the user would see this when hovering or ctrl+click the param in their native IDE so I think might be ok to leave it